core/shared/src/main/scala/cats/effect/IO.scala

```diff
@@ -588,8 +589,20 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
   def onCancel(fin: IO[Unit]): IO[A] =
     IO.OnCancel(this, fin)
 
-  def onError(f: Throwable => IO[Unit]): IO[A] =
-    handleErrorWith(t => f(t).voidError *> IO.raiseError(t))
+  @deprecated("Use onError with PartialFunction argument", "3.6.0")
+  def onError(f: Throwable => IO[Unit]): IO[A] = {
+    val pf: PartialFunction[Throwable, IO[Unit]] = { case t => f(t) }
+    onError(pf)
+  }
+
+  /**
+   * Execute a callback on certain errors, then rethrow them. Any non matching error is rethrown
+   * as well.
+   *
+   * Implements `ApplicativeError.onError`.
+   */
+  def onError(pf: PartialFunction[Throwable, IO[Unit]]): IO[A] =
+    handleErrorWith(t => pf.applyOrElse(t, (_: Throwable) => IO.unit) *> IO.raiseError(t))
```

Comment:
To match the other implementation we need this `voidError`. But this raises an interesting question about the semantics: if the `onError` callback raises an error, which error should be propagated, that error or the original one? Also the `voidError` completely discards the error without reporting it, which seems really weird 😕 

```suggestion
    handleErrorWith(t => pf.applyOrElse(t, (_: Throwable) => IO.unit).voidError *> IO.raiseError(t))
```

Reply:
I think what we should do is extract this into a helper method.
https://github.com/typelevel/cats-effect/blob/7168625ee8f960f6873f2692e73afba75b22d774/core/shared/src/main/scala/cats/effect/IO.scala#L493

Then instead of `voidError` we should redirect errors within the `onError` callback to that? And then re-raise the original error.

Reply:
thanks @armanbilge, I'm completely missed that `voidError`. I added `reportError` helper method as your suggestion. I think this is a great improvement. The name is subjecting for bike-shedding. We can also use this `reportError` method in the two other places. I can do that in this pr.

Reply:
> I'm completely missed that `voidError`

Yes, I almost missed it as well. We should add a test to cover this case (what happens if `onError` callback raises an error), that would have helped us. Actually we might even consider adding this as a law to Cats, so that `onError` semantics are the same across all implementations.

> We can also use this `reportError` method in the two other places. I can do that in this pr.

🚀 it looks good! I think name is reasonable and it is private internal anyway.

Reply:
cool, I'll add a test case for in this first. And think about a law for cats later on.

Reply:
Hmm, actually, now I am hesitating 😅 

The goal of this PR is to make `IO#onError` consistent with `ApplicativeError#onError`. If we use `voidError`/`reportError` then the semantics are inconsistent again with `ApplicativeError` 😕 

https://github.com/typelevel/cats/blob/8e8724a29881a394bc39357a7f0cd21124573f30/core/src/main/scala/cats/ApplicativeError.scala#L261-L262

Unless we change it in Cats (and this may be too drastic of a change) I think we just have to let the callback's error propagate instead.

Reply:
ahhhh, you're right. So, my first implemenaton was accidentally correct :p.